### PR TITLE
[VM Script] Update create instance template logic

### PR DESCRIPTION
**Pull request by jasonwzm**
The PR has the following updates:
- Cluster registration validation
- WorkloadGroup validation
  - ASM revision is labeled correctly
  - Cluster has the correct identity provider
  - Workload is using an existing service account
- Retrieve eastwest gateway IP
- Retrieve cluster network
- Retrieve identity provider

Adding one basic test to test the instance template creation.

To reduce review burden, more tests will be added in following PRs.

### CODE REVIEW

**Comment by jasonwzm** [comment]
@zerobfd WDYT putting VM tests at the same directory level as other tests to better use the common library? Or do you prefer in different directory.

File: scripts/asm-installer/tests/common.sh
Comment (zerobfd) [inline scripts/asm-installer/tests/common.sh:675]: Don't put the vm- prefix here, do it when you call the function. Also there's nothing specific to WORKLOAD here, just call this SERVICE_ACCOUNT (or SA)

File: scripts/asm-installer/tests/common.sh
Comment (zerobfd) [inline scripts/asm-installer/tests/common.sh:710]: Change `return 1` to `false`

File: scripts/asm-installer/tests/run_basic_instance_template_creation
Comment (zerobfd) [inline scripts/asm-installer/tests/run_basic_instance_template_creation:18]: Why are these --enable flags necessary?

File: asm/vm/asm_vm
Comment (zerobfd) [inline asm/vm/asm_vm:237]: It's very non-standard to have a two character flag after a single dash. I don't have a good suggestion for what it should be, but if you can't think of one just remove the single-dash option and make it `--workload_namespace` only.

File: asm/vm/asm_vm
Comment (zerobfd) [inline asm/vm/asm_vm:539]: What does this function name mean?

File: asm/vm/asm_vm
Comment (zerobfd) [inline asm/vm/asm_vm:675]: These double quotes will be printed literally since you're in a heredoc.

File: asm/vm/asm_vm
Comment (zerobfd) [inline asm/vm/asm_vm:895]: You don't break after you find what you're looking for. It looks like you might find multiple UIDs but only set the latest one. Can you return early? Also, is there a reason not to just set CREDENTIAL_IDENTITY_PROVIDER in the same function?

File: scripts/asm-installer/cloudbuild-master.yaml
Comment (zerobfd) [inline scripts/asm-installer/cloudbuild-master.yaml:147]: Can you update the lint rule to also look at asm/vm/asm_vm? Also, since you're now creating Cloud resources, you need to have some sort of logic to clean up ones that leak.

**Review by zerobfd — CHANGES_REQUESTED** [review]

File: scripts/asm-installer/tests/common.sh
Comment (jasonwzm) [reply scripts/asm-installer/tests/common.sh]: > Don't put the vm- prefix here, do it when you call the function. 
I don't understand the benefit here. I don't want to only add it in one place but forget to add it to other places.
> Also there's nothing specific to WORKLOAD here, just call this SERVICE_ACCOUNT (or SA)
This is very specific to workload. Renamed to workload_service_account.

File: scripts/asm-installer/tests/common.sh
Comment (jasonwzm) [reply scripts/asm-installer/tests/common.sh]: Done.

File: scripts/asm-installer/cloudbuild-master.yaml
Comment (jasonwzm) [reply scripts/asm-installer/cloudbuild-master.yaml:148]: Done.

File: asm/vm/asm_vm
Comment (jasonwzm) [reply asm/vm/asm_vm]: Good point. break is added. But it will not have more than one UIDs returned, break is making it more efficient.

> Also, is there a reason not to just set CREDENTIAL_IDENTITY_PROVIDER in the same function?

Not really, just separation of logic. But combined.

File: asm/vm/asm_vm
Comment (jasonwzm) [reply asm/vm/asm_vm]: Removed.

File: asm/vm/asm_vm
Comment (jasonwzm) [reply asm/vm/asm_vm]: It also retrieves the asm revisions. But renamed.

File: asm/vm/asm_vm
Comment (jasonwzm) [reply asm/vm/asm_vm]: Done. Removed.

File: scripts/asm-installer/tests/run_basic_instance_template_creation
Comment (jasonwzm) [reply scripts/asm-installer/tests/run_basic_instance_template_creation]: Yea, i see it is not needed. Removed from VM control plane test as well.

File: scripts/asm-installer/cloudbuild-master.yaml
Comment (jasonwzm) [reply scripts/asm-installer/cloudbuild-master.yaml:148]: > Also, since you're now creating Cloud resources, you need to have some sort of logic to clean up ones that leak.

The test script cleans up leaked objects. Is there somewhere else that also cleans up stuff?

File: scripts/asm-installer/tests/common.sh
Comment (zerobfd) [reply scripts/asm-installer/tests/common.sh]: I mean that there's nothing specific to workloads with the gcloud command you're calling.

If you want to do it like this, I'd rather you have a generic create_service_account() function, and then you can wrap that with create_workload_service_account() where you add the vm- and whatever else. At some point we're going to have to test permissions and create service accounts, and whoever adds that logic is just going to have to pull the logic apart anyways.

I don't care enough to block the PR on it if you don't want to put it in this PR though.

File: scripts/asm-installer/cloudbuild-master.yaml
Comment (zerobfd) [reply scripts/asm-installer/cloudbuild-master.yaml:148]: You're creating service accounts and instance templates in the test and the script might error and close before it reaches the end.

Comment (zerobfd) [reply scripts/asm-installer/cloudbuild-master.yaml:148]: Sorry, I misread your comment at first. It's in the setup_longterm_cluster script. That way things that leak will get cleaned up eventually on the next test run.

File: scripts/asm-installer/tests/run_basic_instance_template_creation
Comment (zerobfd) [reply scripts/asm-installer/tests/run_basic_instance_template_creation]: Thanks!

File: asm/vm/asm_vm
Comment (maasen) [inline asm/vm/asm_vm:49]: Can we shift to referencing Environ instead of Hub?

File: asm/vm/asm_vm
Comment (maasen) [inline asm/vm/asm_vm:239]: nit - space after '(optional)' and possibly ident to match other flags

File: scripts/asm-installer/tests/common.sh
Comment (jasonwzm) [reply scripts/asm-installer/tests/common.sh]: sgtm. I extract create_service_account out for now.

File: scripts/asm-installer/cloudbuild-master.yaml
Comment (jasonwzm) [reply scripts/asm-installer/cloudbuild-master.yaml:148]: Added cleanup all vm instance template and cleanup all workload service accounts to the build step.

File: asm/vm/asm_vm
Comment (jasonwzm) [reply asm/vm/asm_vm]: Done.

File: asm/vm/asm_vm
Comment (jasonwzm) [reply asm/vm/asm_vm]: Done.

File: asm/vm/asm_vm
Comment (zerobfd) [reply asm/vm/asm_vm]: Agree about the rename, but I'd rather not do it here. It needs to be changed in install_asm (and possible other places) as well, and I'd rather have it all in one PR that doesn't change functionality. There's a lot in this PR already.

File: asm/vm/asm_vm
Comment (zerobfd) [inline asm/vm/asm_vm:49]: Please change back to HUB for this PR re: other comment thread

**Review by zerobfd — APPROVED** [review]
LGTM as soon as tests pass if the only thing left are test fixes. Re-request review if the logic materially changes.

File: asm/vm/asm_vm
Comment (jasonwzm) [reply asm/vm/asm_vm]: I am removing this membership id variable. This is not needed by VM script.

**Pull request merged**
